service/init.go

```diff
@@ -331,19 +355,58 @@ func (pcf *PCF) registerNF() {
 			}
 			initLog.Infof("Pcf Profile Registering to NRF: %v", profile)
 			//Indefinite attempt to register until success
-			_, self.NfId, err = consumer.SendRegisterNFInstance(self.NrfUri, self.NfId, profile)
+			profile, _, self.NfId, err = consumer.SendRegisterNFInstance(self.NrfUri, self.NfId, profile)
 			if err != nil {
 				initLog.Errorf("PCF register to NRF Error[%s]", err.Error())
 			} else {
+				pcf.StartKeepAliveTimer(profile)
 				//NRF Registration Successful, Trigger for UDR Discovery
-				pcf.discoverUdr()
+				pcf.DiscoverUdr()
+			}
+		} else {
+			//stopping keepAlive timer
+			pcf.StopKeepAliveTimer()
+			initLog.Infof("PCF is not having Minimum Config to Register/Update to NRF")
+			problemDetails, err := consumer.SendDeregisterNFInstance()
+			if problemDetails != nil {
+				initLog.Errorf("PCF Deregister Instance to NRF failed, Problem: [+%v]", problemDetails)
+			}
+			if err != nil {
+				initLog.Errorf("PCF Deregister Instance to NRF Error[%s]", err.Error())
+			} else {
+				logger.InitLog.Infof("Deregister from NRF successfully")
 			}
 		}
 	}
+}
+
+//UpdateNF is the callback function, this is called when keepalivetimer elapsed
+func (pcf *PCF) UpdateNF() {
+	//self := context.PCF_Self()
+	pitem := models.PatchItem{
+		Op:    "replace",
+		Path:  "/nfStatus",
+		Value: "REGISTERED",
+	}
+	var patchItem []models.PatchItem
+	patchItem = append(patchItem, pitem)
+	nfProfile, problemDetails, err := consumer.SendUpdateNFInstance(patchItem)
+	if err != nil {
+		initLog.Errorf("PCF update to NRF Error[%s]", err.Error())
+	} else if problemDetails != nil {
+		err = fmt.Errorf("PCF update to NRF ProblemDetails[%v]", problemDetails)
+	}
 
+	if nfProfile.HeartBeatTimer == 0 {
```

Comment:
In failure case, nfProfile shall be nil. Add failure handling

Reply:
done